server/comms/server.go

```diff
@@ -452,3 +553,39 @@ func parseListeners(addrs []string) ([]string, []string, bool, error) {
 	}
 	return ipv4ListenAddrs, ipv6ListenAddrs, haveWildcard, nil
 }
+
+// handleHTTP handles an HTTP request by finding and calling the appropriate
+// handler. The caller of handleHTTP should have already parsed and request
+// data into the right type for the handler.
+func handleHTTP(w http.ResponseWriter, r *http.Request, route string, thing interface{}) {
+	handler := httpRoutes[route]
+	if handler == nil {
+		http.Error(w, http.StatusText(http.StatusNotFound), http.StatusNotFound)
+		return
+	}
+	resp, err := handler(thing)
+	if err != nil {
+		writeJSONWithStatus(w, map[string]string{"error": err.Error()}, http.StatusBadRequest)
+		return
+	}
+	writeJSONWithStatus(w, resp, http.StatusOK)
+}
+
+// routeHandler creates a HandlerFunc for a route. Middleware should have
+// already processed the request and added the request struct to the Context.
+func routeHandler(route string) func(w http.ResponseWriter, r *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		handleHTTP(w, r, route, r.Context().Value(ctxThing))
```

Comment:
It's unfortunate that each request will do a lookup in the `httpRoutes` map.  I was wondering if this could be:

```go
// routeHandler creates a HandlerFunc for a route. Middleware should have
// already processed the request and added the request struct to the Context.
func routeHandler(route string) func(w http.ResponseWriter, r *http.Request) {
	handler := httpRoutes[route]
	if handler == nil {
		panic("no known handler")
	}
	return func(w http.ResponseWriter, r *http.Request) {
		thing := r.Context().Value(ctxThing)
		resp, err := handler(thing)
		if err != nil {
			writeJSONWithStatus(w, map[string]string{"error": err.Error()}, http.StatusBadRequest)
			return
		}
		writeJSONWithStatus(w, resp, http.StatusOK)
	}
}
```

Or might all the http routes not be registered when comms.Server starts?

Reply:
They actually weren't, but that was breaking my own rules from the docs of `(*Server).Run`, which said

> `Run should be called only after all routes are registered`

 so I've corrected the sequence in **server/dex** and used your suggestion here.